libdnf-plugins/actions/actions.cpp

```diff
@@ -0,0 +1,725 @@
+
+    // Parsed actions for individual hooks
+    std::vector<Action> pre_base_setup_actions;
+    std::vector<Action> post_base_setup_actions;
+    std::vector<Action> pre_trans_actions;
+    std::vector<Action> post_trans_actions;
+
+    // cache for sharing between pre_transaction and post_transaction hooks
+    bool transaction_cached = false;
+    std::vector<libdnf::base::TransactionPackage> trans_packages;
+    std::map<libdnf::rpm::PackageId, const libdnf::base::TransactionPackage *> pkg_id2trans_pkg;
+    std::optional<libdnf::rpm::PackageQuery> in_full_query;
+    std::optional<libdnf::rpm::PackageQuery> out_full_query;
+    std::optional<libdnf::rpm::PackageQuery> all_full_query;
+
+    // store temporaly variables for sharing data between actions (executables)
+    std::map<std::string, std::string> tmp_variables;
+};
+
+
+class ActionsPluginError : public libdnf::Error {
+    using Error::Error;
+    const char * get_domain_name() const noexcept override { return "libdnf::plugin"; }
+    const char * get_name() const noexcept override { return "ActionsPluginError"; }
+};
+
+
+bool CommandToRun::operator<(const CommandToRun & other) const noexcept {
+    if (command == other.command) {
+        if (args.size() == other.args.size()) {
+            for (size_t i = 0; i < args.size(); ++i) {
+                if (args[i] != other.args[i]) {
+                    return args[i] < other.args[i];
+                }
+            }
+        }
+        return args.size() < other.args.size();
+    }
+    return command < other.command;
+}
+
+// splits the string into a vector, delimiter is space, spaces escaped by \ are ignored
+std::vector<std::string> split(const std::string & str) {
+    std::vector<std::string> ret;
+
+    bool escape = false;
+    auto it = str.begin();
+    while (true) {
+        while (*it == ' ') {
+            ++it;
+        }
+        if (*it == '\0' || *it == '\n' || *it == '\r') {
+            break;
+        }
+        auto it_start = it;
+        for (; (escape || *it != ' ') && *it != '\0' && *it != '\n' && *it != '\r'; ++it) {
+            escape = !escape && *it == '\\';
+        }
+        ret.emplace_back(it_start, it);
+    }
+
+    return ret;
+}
+
+std::pair<std::string, bool> Actions::substitute(
+    const libdnf::base::TransactionPackage * trans_pkg,
+    const libdnf::rpm::Package * pkg,
+    std::string_view in,
+    std::filesystem::path file,
+    int line_number) {
+    auto & logger = *base.get_logger();
+    std::string ret;
+    bool error = false;
+    size_t pos = 0;
+    do {
+        auto var_pos = in.find("${", pos);
+        ret += in.substr(pos, var_pos - pos);
+        if (var_pos == std::string::npos) {
+            break;
+        }
+        var_pos += 2;
+        auto var_end_pos = in.find('}', var_pos);
+        if (var_end_pos == std::string::npos) {
+            logger.error(
+                "Actions plugin: Syntax error: Incomplete variable name \"{}\" in file \"{}\" at line {}",
```

Comment:
Please use `"... on line {}"` instead of "at line", repeats multiple times throughout the file.

Reply:
OK